### PR TITLE
avm2: Add AS3 native initializers, abstract class metadata, remove separate super_init method entirely

### PR DESCRIPTION
**Pull request by adrian17**
Requires #18358.

### CODE REVIEW

File: core/build_playerglobal/src/lib.rs [hunk 1]
```diff
@@ -36,6 +36,9 @@ const METADATA_CALL_HANDLER: &str = "CallHandler";
 /// Indicates that we should generate a reference to a custom constructor
 /// method (used as a metadata key with `Ruffle` metadata)
 const METADATA_CUSTOM_CONSTRUCTOR: &str = "CustomConstructor";
+/// Indicates that the class can't be directly instantiated (but its child classes might be).
+/// Binds to an always-throwing allocator.
+const METADATA_ABSTRACT: &str = "Abstract";
```
Comment (Lord-McSweeney) [inline core/build_playerglobal/src/lib.rs:38]: This should also definitely be more specific

**Comment by Lord-McSweeney** [comment]
`Worker` and `MorphShape` can also be `Abstract` (unless you want to leave that to a follow-up, since there are several other classes that should be)

File: core/build_playerglobal/src/lib.rs [hunk 2]
```diff
@@ -403,28 +406,70 @@ fn write_native_table(data: &[u8], out_dir: &Path) -> Result<Vec<u8>, Box<dyn st
...
         let class_name = resolve_multiname_name(
             &abc,
-            &abc.constant_pool.multinames[class_name_idx as usize - 1],
+            &abc.constant_pool.multinames[class_name_idx.0 as usize - 1],
         );
 
         let instance_allocator_method_name =
             "::".to_string() + &flash_to_rust_path(&class_name) + "_allocator";
         let super_init_method_name = "::super_init".to_string();
+        let init_method_name = "::init".to_string();
```
[the remark this answers is not in the document]
Comment (adrian17) [reply core/build_playerglobal/src/lib.rs]: I already renamed this in a later commit, to `<class_name>_initializer`, after noticing that we already have a ton of functions named `init` :)

**Comment by adrian17** [comment]
> unless you want to leave that to a follow-up

I only modified the classes that I already needed to touch during the refactor, that's why I didn't touch Worker and MorphShape - but sure :) Will add them to the test too.

**Review by Lord-McSweeney — APPROVED** [review]

**Pull request merged**